src/v/s3/signature.cc

```diff
@@ -0,0 +1,399 @@
+static result<std::string> get_canonical_query_string(std::string target) {
+    if (target.empty() || target[0] != '/') {
+        vlog(s3_log.error, "invalid URI {}", target);
+        return make_error_code(s3_error_codes::invalid_uri);
+    }
+    auto pos = target.find('?');
+    if (pos == std::string::npos || pos == target.size() - 1) {
+        return "";
+    }
+    auto query_str = target.substr(pos + 1);
+    std::vector<std::string> params;
+    boost::split(params, query_str, boost::is_any_of("&"));
+    std::vector<std::pair<std::string, std::string>> query_params;
+    for (auto const& param : params) {
+        auto p = param.find('=');
+        if (p == absl::string_view::npos) {
+            // parameter with empty value
+            query_params.emplace_back(param, "");
+        } else {
+            if (p == 0) {
+                // parameter value can be empty but name can't
+                return make_error_code(s3_error_codes::invalid_uri_params);
+            }
+            std::string pname = param.substr(0, p);
+            std::string pvalue = param.substr(p + 1);
+            query_params.emplace_back(pname, pvalue);
+        }
+    }
+    std::sort(query_params.begin(), query_params.end());
+
+    // Generate canonical query string
+    std::string result;
+    int cnt = 0;
+    for (auto [pname, pvalue] : query_params) {
+        if (cnt++ > 0) {
+            result.push_back('&');
+        }
+        result += uri_encode(pname, true);
+        result += "=";
+        result += uri_encode(pvalue, true);
+    }
+    return result;
+}
+
+struct canonical_headers {
+    std::string canonical_headers; // string that contains canonical headers
+    std::string signed_headers;    // string that contains list of header names
+};
+
+/// CanonicalHeaders is a list of request headers with their values.
+/// Individual header name and value pairs are separated by the newline
+/// character ("\n"). Header names must be in lowercase. You must sort the
+/// header names alphabetically to construct the string, as shown in the
+/// following example:
+///
+///   Lowercase(<HeaderName1>)+":"+Trim(<value>)+"\n"
+///   Lowercase(<HeaderName2>)+":"+Trim(<value>)+"\n"
+///   ...
+///   Lowercase(<HeaderNameN>)+":"+Trim(<value>)+"\n"
+///
+/// The CanonicalHeaders list must include the following:
+///
+///   - HTTP host header.
+///   - If the Content-Type header is present in the request, you must add it to
+///   the
+///     CanonicalHeaders list.
+///   - Any x-amz-* headers that you plan to include in your request must also
+///   be added.
+///     For example, if you are using temporary security credentials, you need
+///     to include x-amz-security-token in your request. You must add this
+///     header in the list of CanonicalHeaders.
+///
+///
+static result<canonical_headers>
+get_canonical_headers(const http::client::request_header& request) {
+    std::vector<std::pair<std::string, std::string>> headers;
```

Comment:
i'd use an anonymous struct with an `string operator()` conversion operator that tdoes the concatenation.